CONTRIBUTING.md

```diff
@@ -27,8 +27,90 @@ Realm welcomes all contributions! The only requirement we have is that, like man
 
 ## Unit tests
 
-All PR's must be accompanied by related unit tests.
-
+All PR's must be accompanied by related unit tests. All bug fixes must have a unit test proving that the bug is fixed.
+You can use `./realm/gradlew connectedCheck createDebugCoverageReport` to generate a coverage report to check for 
+missing unit test coverage. The aim is 100% code coverage.
+
+When writing unit tests, use the following guide lines:
+
+1) Unit tests must be written using JUnit4.
+
+2) All tests for a class should be grouped in a class called `<className>Tests`, unless the functionality is cross-
+   cutting like [`RxJavaTests`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/RxJavaTests.java) 
+   or [`RealmAsyncQueryTests`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java).
+
+3) Test methods should use camelCase and underscore `_` between logical sections to increase method name readability. 
+   Methods should ideally start with the name of the method being tested. Patterns like: `<methodName>_<description>`, 
+   `<methodName>_<param>_<description>` or `<description>` are encouraged.
+   
+4) All unit tests creating Realms must do so using the [`TestRealmConfigurationFactory`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/rule/TestRealmConfigurationFactory.java) 
+   or [`RunInLooperThread`](https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java) 
+   test rules. This ensures that all Realms are properly closed and deleted between each test.
+
+5) Use the `@RunInLooperThread` rule for any test that depends on Realms notification system. 
+
+6) Input-parameters should be boundary tested. Especially `Null/NotNull`, but also the state of Realm objects like
+   standalone objects, deleted objects, objects from other threads.
+
+7) Unit tests are not required to only have 1 test. It is acceptable to combine multiple tests into one unit test, but
+   if it fails, it should be clear why it failed.
+
+8) Use `@Test(expected = xxx.class)` or the `ExceptedException` rule to detect exceptions.
```

Comment:
I would strongly recommend against using this. It only works as intended for one line methods. Otherwise you have no control over when the exception is supposed to be thrown. We used to use it back in the days, but went away from it for that reason.